### PR TITLE
Add support for auto-activated keyed singletons

### PR DESCRIPTION
**Pull request by geeknoid**
null

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/extensions/pull/4222)

### CODE REVIEW

**Comment by geeknoid** [comment]
Tests will fail in this PR until https://github.com/dotnet/runtime/issues/89447 is fixed.

File: test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs [hunk 1]
```diff
@@ -0,0 +1,367 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
```
Comment (xakep139) [inline test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs:5]:
```suggestion
```

**Comment by RussKie** [comment]
We're blocked by https://github.com/dotnet/aspnetcore/pull/49704.

**Comment by RussKie** [comment]
Microsoft.Extensions.DependencyInjection.AutoActivation.Tests are failing

**Comment by xakep139** [comment]
https://github.com/dotnet/aspnetcore/pull/49704 was merged 14 hours ago, I will try to update the repo.

**Comment by RussKie** [comment]
Hopefully https://github.com/dotnet/extensions/pull/4240 brought the latest changes.

**Comment by xakep139** [comment]
@geeknoid can you please try to merge main into this branch and see whether it helps?

**Comment by geeknoid** [comment]
> @geeknoid can you please try to merge main into this branch and see whether it helps?

Yep, the fundamental issue is fixed. Now I just have to get my own shitty code working :-)

File: src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+            return;
+        }
+
+        services.Add(descriptor);
+        _ = services.ActivateKeyed<TService>(descriptor.ServiceKey);
+    }
+
+    private static void TryAddAndActivateKeyed(this IServiceCollection services, ServiceDescriptor descriptor)
+    {
+        if (services.Any(d => d.ServiceType == descriptor.ServiceType))
```
Comment (xakep139) [inline src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs:397]:
```suggestion
        if (services.Any(d => d.ServiceType == descriptor.ServiceType && d.ServiceKey == descriptor.ServiceKey))
```

File: src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+        _ = Throw.IfNull(services);
+        _ = Throw.IfNull(implementationFactory);
+
+        services.TryAddAndActivateKeyed<TService>(ServiceDescriptor.KeyedSingleton<TService>(serviceKey, implementationFactory));
+    }
+
+    private static void TryAddAndActivateKeyed<TService>(this IServiceCollection services, ServiceDescriptor descriptor)
+        where TService : class
+    {
+        if (services.Any(d => d.ServiceType == descriptor.ServiceType))
```
Comment (xakep139) [inline src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs:386]:
```suggestion
        if (services.Any(d => d.ServiceType == descriptor.ServiceType && d.ServiceKey == descriptor.ServiceKey))
```

File: src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs (hunk 2 above)
Comment (geeknoid) [reply src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs]: Done

File: src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs (hunk 3 above)
Comment (geeknoid) [reply src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs]: Done.

**Comment by geeknoid** [comment]
@xakep139 This PR is finally ready to go...

File: src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+{
+    /// <summary>
+    /// Enforces singleton activation at startup time rather then at runtime.
+    /// </summary>
+    /// <typeparam name="TService">The type of the service to activate.</typeparam>
+    /// <param name="services">The <see cref="IServiceCollection"/> to add the service to.</param>
+    /// <param name="serviceKey">The <see cref="ServiceDescriptor.ServiceKey"/> of the service.</param>
+    /// <returns>A reference to this instance after the operation has completed.</returns>
+    [Experimental(diagnosticId: Experiments.AutoActivation, UrlFormat = Experiments.UrlFormat)]
+    public static IServiceCollection ActivateKeyed<TService>(
```
Comment (xakep139) [inline src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs:23]: Should we rename it to `ActivateKeyedSingleton`? All DI-related extension method specify desired lifetime it their name

File: src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+                }
+
+                _ = ao.KeyedAutoActivators.Add((serviceType, serviceKey));
+            });
+
+        return services;
+    }
+
+    /// <summary>
+    /// Adds an auto-activated singleton service of the type specified in TService with an implementation
```
Comment (xakep139) [inline src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs:97]:
```suggestion
    /// Adds an auto-activated singleton service of the type specified in <c>TService</c> with an implementation
```

File: src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+    /// type specified in TImplementation using the factory specified in implementationFactory
```
Comment (xakep139) [inline src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs:98]:
```suggestion
    /// type specified in <c>TImplementation</c> using the factory specified in <paramref name="implementationFactory"/>
```

File: src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+        where TService : class
+        where TImplementation : class, TService
+    {
+        _ = Throw.IfNull(services);
+
+        services.TryAddAndActivateKeyed<TService>(ServiceDescriptor.KeyedSingleton<TService, TImplementation>(serviceKey));
+    }
+
+    /// <summary>
+    /// Adds an auto-activated singleton service of the type specified in serviceType with a factory
```
Comment (xakep139) [inline src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs:362]: There's no `serviceType` in this overload

File: src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs (hunk 4 above)
Comment (xakep139) [reply src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs:23]: Alternatively, if we don't care about lifetime, we should remove `singleton` from XML docs

File: src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+        _ = Throw.IfNull(services);
+        _ = Throw.IfNull(implementationFactory);
+
+        return services
+            .AddKeyedSingleton<TService, TImplementation>(serviceKey, implementationFactory)
+            .ActivateKeyed<TService>(serviceKey);
+    }
+
+    /// <summary>
+    /// Adds an auto-activated singleton service of the type specified in TService with an implementation
```
Comment (xakep139) [inline src/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation/AutoActivationExtensions.Keyed.cs:124]:
```suggestion
    /// Adds an auto-activated keyed singleton service of the type specified in TService with an implementation
```

File: test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs [hunk 9]
```diff
@@ -0,0 +1,425 @@
...
+        Assert.Equal(0, instanceCount.Counter);
+
+        var serviceKey = new object();
+        using var host = await FakeHost.CreateBuilder()
+            .ConfigureServices(services => services
+                .AddSingleton<IFakeServiceCounter>(instanceCount)
+                .AddActivatedKeyedSingleton<IFakeService, FakeService>(serviceKey))
+            .StartAsync();
+
+        var service = host.Services.GetKeyedService<IFakeService>(serviceKey);
```
Comment (xakep139) [inline test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs:30]: What if we also check that `instanceCount.Counter` equals to `1` before we get the service from the provider?

File: test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+        var childCount = new InstanceCreatingCounter();
+
+        var serviceKey = new object();
+        using var host = await FakeHost.CreateBuilder()
+            .ConfigureServices(services => services.AddSingleton<IFakeServiceCounter>(childCount)
+                .AddSingleton<IFactoryServiceCounter>(parentCount)
+                .AddActivatedKeyedSingleton(typeof(IFakeService), serviceKey, typeof(FakeService))
+                .AddActivatedKeyedSingleton<IFactoryService, FactoryService>(serviceKey, (sp, sk) =>
+                {
+                    return new FactoryService(sp.GetKeyedService<IFakeService>(sk)!, sp.GetRequiredService<IFactoryServiceCounter>()!);
```
Comment (xakep139) [inline test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs:50]:
```suggestion
                    return new FactoryService(sp.GetKeyedService<IFakeService>(sk)!, sp.GetRequiredService<IFactoryServiceCounter>());
```

File: test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs [hunk 11]
```diff
@@ -0,0 +1,425 @@
...
+
+        var service = host.Services.GetKeyedService<IFakeService>(serviceKey);
+        await host.StopAsync();
+
+        Assert.NotNull(service);
+        Assert.Equal(1, instanceCount.Counter);
+    }
+
+    [Fact]
+    public async Task ShouldAddAndActivateOnlyOnce_WhenHasChildAsync_Keyed()
```
Comment (xakep139) [inline test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs:38]:
```suggestion
    public async Task ShouldAddAndActivateOnlyOnce_WhenHasChild_Keyed()
```

File: test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs [hunk 12]
```diff
@@ -0,0 +1,425 @@
...
+                }))
+            .StartAsync();
+        await host.StopAsync();
+
+        Assert.Equal(1, childCount.Counter);
+        Assert.Equal(1, parentCount.Counter);
+    }
+
+    [Fact]
+    public async Task ShouldResolveComponentsAutomaticallyAsync_Keyed()
```
Comment (xakep139) [inline test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs:60]:
```suggestion
    public async Task ShouldResolveComponentsAutomatically_Keyed()
```

File: test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs [hunk 13]
```diff
@@ -0,0 +1,425 @@
...
+                .AddActivatedKeyedSingleton<IFactoryService, FactoryService>(serviceKey))
+            .StartAsync();
+        await host.StopAsync();
+
+        Assert.Equal(1, childCount.Counter);
+        Assert.Equal(1, parentCount.Counter);
+    }
+
+    [Fact]
+    public async Task CanActivateEnumerableAsync_Keyed()
```
Comment (xakep139) [inline test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs:80]:
```suggestion
    public async Task CanActivateEnumerable_Keyed()
```

File: test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs [hunk 14]
```diff
@@ -0,0 +1,425 @@
...
+                services.TryAddActivatedKeyedSingleton<IFakeService, AnotherFakeService>(serviceKey);
+            })
+            .StartAsync();
+        await host.StopAsync();
+
+        Assert.Equal(1, counter.Counter);
+        Assert.Equal(0, anotherFakeServiceCount.Counter);
+    }
+
+    // ------------------------------------------------------------------------------
```
Comment (xakep139) [inline test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs:303]: Remove that line?

**Review by xakep139 — APPROVED** [review]

**Pull request merged**